app/src/main/java/de/westnordost/streetcomplete/quests/railway_crossing/AddRailwayCrossingBarrier.kt

```diff
@@ -1,21 +1,50 @@
 package de.westnordost.streetcomplete.quests.railway_crossing
 
+import de.westnordost.osmapi.map.data.BoundingBox
+import de.westnordost.osmapi.map.data.Element
 import de.westnordost.streetcomplete.R
-import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType
+import de.westnordost.streetcomplete.data.osm.OsmElementQuestType
 import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder
+import de.westnordost.streetcomplete.data.osm.download.MapDataWithGeometryHandler
 import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao
+import de.westnordost.streetcomplete.data.osm.tql.OverpassQLUtil
+import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment
 
-class AddRailwayCrossingBarrier(o: OverpassMapDataDao) : SimpleOverpassQuestType<String>(o) {
+class AddRailwayCrossingBarrier(private val overpassMapDataDao: OverpassMapDataDao) : OsmElementQuestType<RailwayCrossingBarrierAnswer> {
+    override fun createForm(): AbstractQuestAnswerFragment<RailwayCrossingBarrierAnswer> {
+        return AddRailwayCrossingBarrierForm()
+    }
 
-    override val tagFilters = "nodes with railway=level_crossing and !crossing:barrier"
-    override val commitMessage = "Add type of barrier for railway crossing"
-    override val icon = R.drawable.ic_quest_railway
+    override fun download(bbox: BoundingBox, handler: MapDataWithGeometryHandler): Boolean {
+        return overpassMapDataDao.getAndHandleQuota(getOverpassQuery(bbox), handler);
+    }
 
-    override fun getTitle(tags: Map<String, String>) = R.string.quest_railway_crossing_barrier_title
+    override fun isApplicableTo(element: Element): Boolean? {
+        return null
+    }
 
-    override fun createForm() = AddRailwayCrossingBarrierForm()
+    override fun applyAnswerTo(answer: RailwayCrossingBarrierAnswer, changes: StringMapChangesBuilder) {
+        when {
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.MISSING -> changes.add("crossing:barrier", "no")
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.HALF -> changes.add("crossing:barrier", "half")
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.DOUBLE_HALF -> changes.add("crossing:barrier", "double_half")
+            answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.FULL -> changes.add("crossing:barrier", "full")
+        }
+    }
 
-    override fun applyAnswerTo(answer: String, changes: StringMapChangesBuilder) {
-        changes.add("crossing:barrier", answer)
```

Comment:
Actually, there is no reason to change anything here from the current behavior. 

But anyway, here is another Kotlin tip:
`if` and `when` statements are itself also statements, so you can write
`val x = if(a) 1 else if(c) 2 else 3`

So, here you could refactor this method to look like this:

```kotlin
override fun applyAnswerTo(answer: RailwayCrossingBarrierAnswer, changes: StringMapChangesBuilder) {
    changes.add("crossing:barrier", when {
        answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.MISSING -> "no"
        answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.HALF -> "half"
        answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.DOUBLE_HALF -> "double_half"
        answer.barrierType == AddRailwayCrossingBarrierForm.BarrierOption.FULL -> "full"
    })
}
```
And since `when` is just a super-powered Java `case`, you can further simplify to
```kotlin
override fun applyAnswerTo(answer: RailwayCrossingBarrierAnswer, changes: StringMapChangesBuilder) {
    changes.add("crossing:barrier", when(answer.barrierType) {
        AddRailwayCrossingBarrierForm.BarrierOption.MISSING -> "no"
        AddRailwayCrossingBarrierForm.BarrierOption.HALF -> "half"
        AddRailwayCrossingBarrierForm.BarrierOption.DOUBLE_HALF -> "double_half"
        AddRailwayCrossingBarrierForm.BarrierOption.FULL -> "full"
    })
}
```

But I am just telling you this because I assume that you may be interested in learning more Kotlin. In case not, ignore everything of this comment except my first sentence ;-)